onnxruntime/core/mlas/lib/q4gemm_neon.cpp

```diff
@@ -0,0 +1,803 @@
+
+                            int8x16_t bv_bytes[2];
+                            bv_bytes[0] = vandq_s8(bv_packed, LowMask);
+                            bv_bytes[1] = vandq_s8(vshrq_n_s8(bv_packed, 4), LowMask);
+
+                            // dequantize B
+
+                            // subtract zero point
+                            const int8x16_t zpv = vdupq_n_s8(zp[nn]);
+
+                            bv_bytes[0] = vsubq_s8(bv_bytes[0], zpv);
+                            bv_bytes[1] = vsubq_s8(bv_bytes[1], zpv);
+
+                            // widen to int16
+                            int16x8_t bv_int16[4];
+                            bv_int16[0] = vmovl_s8(vget_low_s8(bv_bytes[0]));
+                            bv_int16[1] = vmovl_s8(vget_high_s8(bv_bytes[0]));
+                            bv_int16[2] = vmovl_s8(vget_low_s8(bv_bytes[1]));
+                            bv_int16[3] = vmovl_s8(vget_high_s8(bv_bytes[1]));
+
+                            // 32 floats of B
+                            float32x4_t bv[8];
+
+                            // widen to int32, cast to float32
+                            bv[0] = vcvtq_f32_s32(vmovl_s16(vget_low_s16(bv_int16[0])));
+                            bv[1] = vcvtq_f32_s32(vmovl_s16(vget_high_s16(bv_int16[0])));
+
+                            bv[2] = vcvtq_f32_s32(vmovl_s16(vget_low_s16(bv_int16[1])));
+                            bv[3] = vcvtq_f32_s32(vmovl_s16(vget_high_s16(bv_int16[1])));
+
+                            bv[4] = vcvtq_f32_s32(vmovl_s16(vget_low_s16(bv_int16[2])));
+                            bv[5] = vcvtq_f32_s32(vmovl_s16(vget_high_s16(bv_int16[2])));
+
+                            bv[6] = vcvtq_f32_s32(vmovl_s16(vget_low_s16(bv_int16[3])));
+                            bv[7] = vcvtq_f32_s32(vmovl_s16(vget_high_s16(bv_int16[3])));
+
+                            // multiply by scale
+                            q4gemm_neon::UnrolledLoop<8>(
+                                [&](size_t j) { bv[j] = vmulq_n_f32(bv[j], scale[nn]); });
+
+                            // c += a * b
+                            q4gemm_neon::UnrolledLoop<8>(
+                                [&](size_t j) { acc[nn] = vfmaq_f32(acc[nn], av[j], bv[j]); });
+
+                            // increment b data pointers to next 32 elements
+                            b_data[nn] += 16;
+                        }
+                    }
+
+                    PackedBBlobPtr += Q4Type::BlobSize;
+                }
+
+                for (int64_t nn = 0; nn < nblk; ++nn) {
+                    SumPtr[nn] = vaddvq_f32(acc[nn]);
+                    SumPtr[nn] += Bias != nullptr ? BiasPtr[nn] : 0.0f;
+                }
+            }
+
+            ARowPtr += lda;
+            CRowPtr += ldc;
+        }
+
+        return CountM;
+    };
+
+    // return impl0_reference();
+    return impl3_four_cols();
+    // return impl5_four_cols_inline();
```

Comment:
note: impl3 and impl5 were pretty close in my measurements of the benchmark Q4GEMM/Q4Sym/M:1/N:4096/K:4096/Threads:8/real_time.

real time, median of 20:
impl3: 1101799 ns
impl5: 1095466 ns

impl3 is easier to maintain.